modules/util/slice.go

```diff
@@ -1,17 +1,80 @@
 // Copyright 2022 The Gitea Authors. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+// Most of the functions in this file can have better implementations with "golang.org/x/exp/slices".
+// However, "golang.org/x/exp" is experimental and unreliable, we shouldn't use it.
+// So lets waiting for the "slices" has be promoted to the main repository one day.
+
 package util
 
-// RemoveIDFromList removes the given ID from the slice, if found.
-// It does not preserve order, and assumes the ID is unique.
-func RemoveIDFromList(list []int64, id int64) ([]int64, bool) {
-	n := len(list) - 1
-	for i, item := range list {
-		if item == id {
-			list[i] = list[n]
-			return list[:n], true
+import "strings"
+
+// IsInSlice returns true if the target exists in the slice.
+func IsInSlice[T comparable](target T, slice []T) bool {
+	return IsInSliceFunc(func(t T) bool { return t == target }, slice)
+}
+
+// IsInSliceFunc returns true if any element in the slice satisfies the targetFunc.
+func IsInSliceFunc[T comparable](targetFunc func(T) bool, slice []T) bool {
+	for _, v := range slice {
+		if targetFunc(v) {
+			return true
+		}
+	}
+	return false
+}
+
+// IsStringInSlice sequential searches if string exists in slice.
+func IsStringInSlice(target string, slice []string, insensitive ...bool) bool {
+	if len(insensitive) != 0 && insensitive[0] {
+		target = strings.ToLower(target)
+		return IsInSliceFunc(func(t string) bool { return strings.ToLower(t) == target }, slice)
+	}
+
+	return IsInSlice(target, slice)
+}
+
+// IsEqualSlice returns true if slices are equal.
+// Be careful, two slice which have the same elements but different order are considered equal here.
+func IsEqualSlice[T comparable](target, source []T) bool {
+	if len(target) != len(source) {
+		return false
+	}
+
+	if (target == nil) != (source == nil) {
```

Comment:
```suggestion
	if target == nil || source == nil {
```

Reply:
What do you mean `IsEmptySlice`? I don't see the function and I don't think it's necessary to add one like `len(s) == 0`.

If we agree that `nil` and `[]` are equal slices, I think we can just remove the check:

```diff
-	if (target == nil) != (source == nil) {
-		return false
-	}
```

Reply:
No, we can't, otherwise it will panic. Use `len(s) == 0` is better than `IsEmptySlice`.

Reply:
What do you mean it will panic? Actually, I did just remove it and added tests, and it   looks well.